# [LAB-05] 2. 연습문제 - Housing Dataset 품질 검사

## #01. 준비작업

### 1. 필요한 라이브러리 참조

In [1]:
from jussam import load_data
from helpers import my_qtcheck

📦 연세대학교 주영아 교수가 제작한 라이브러리를 사용중입니다.
📧 Email(1): j.purplerose@yonsei.ac.kr
📧 Email(2): j.purplerose@gmail.com
📝 Website: https://juyounga.kr/


### 2. 데이터 불러오기

In [2]:
origin = load_data("housing")
origin.head()

📚 주택 가격 데이터셋. 주택 면적, 침실 수, 가구 비치 여부, 주요 도로와의 근접성 등 특정 요소를 기반으로 주택 가격을 예측하는 간단하면서도 도전적인 프로젝트입니다. 데이터셋은 작지만, 다중공선성이 매우 높아 예측 모델 구축에 어려움이 있습니다. 이러한 난관을 극복하고 괜찮은 예측 모델을 만들 수 있을까요? (출처: https://www.kaggle.com/datasets/yasserh/housing-prices-dataset)

    field             description
--  ----------------  ----------------------------------------------------------------------------------------------------------------------------------------------------------------
 0  price             주택 가격 (USD 달러) - 예측 대상 변수
 1  area              주택 면적 (제곱 단위) - 주택의 건축 면적
 2  bedrooms          침실 수 (개수) - 주택의 침실 개수
 3  bathrooms         욕실 수 (개수) - 주택의 욕실 개수
 4  stories           주택 층 수 (개수) - 주택의 총 지상 층수
 5  mainroad          주요 도로 인접 여부 (이진형: Yes/No 또는 0/1) - 주택이 주요 도로에 인접해 있는지 여부
 6  guestroom         게스트룸 보유 여부 (이진형: Yes/No 또는 0/1) - 손님용 방이 있는지 여부
 7  basement          지하실 보유 여부 (이진형: Yes/No 또는 0/1) - 지하실이 있는지 여부
 8  hotwaterheating   온수 난방 시스템 보유 여부 (이진형: Yes/No 또는 0/1) - 온수 난방 시스템이 설치되어 있는지 여부
 9  a

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished


## #02. 자료형 점검

### 1. 자료형 확인

In [3]:
origin.info()

<class 'pandas.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   price             545 non-null    int64
 1   area              545 non-null    int64
 2   bedrooms          545 non-null    int64
 3   bathrooms         545 non-null    int64
 4   stories           545 non-null    int64
 5   mainroad          545 non-null    str  
 6   guestroom         545 non-null    str  
 7   basement          545 non-null    str  
 8   hotwaterheating   545 non-null    str  
 9   airconditioning   545 non-null    str  
 10  parking           545 non-null    int64
 11  prefarea          545 non-null    str  
 12  furnishingstatus  545 non-null    str  
dtypes: int64(6), str(7)
memory usage: 69.2 KB


### 2. 자료형 변환

In [4]:
df1 = my_qtcheck.set_type(origin, as_category=['mainroad', 'guestroom', 'basement', 'hotwaterheating', 'airconditioning', 'prefarea', 'furnishingstatus'])
df1.head()

<class 'pandas.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    int64   
 1   area              545 non-null    int64   
 2   bedrooms          545 non-null    int64   
 3   bathrooms         545 non-null    int64   
 4   stories           545 non-null    int64   
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    category
 7   basement          545 non-null    category
 8   hotwaterheating   545 non-null    category
 9   airconditioning   545 non-null    category
 10  parking           545 non-null    int64   
 11  prefarea          545 non-null    category
 12  furnishingstatus  545 non-null    category
dtypes: category(7), int64(6)
memory usage: 29.6 KB


,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished


## #03. 데이터 중복 점검

### 1. 중복 검사 및 중복 데이터 삭제

In [5]:
df2 = my_qtcheck.check_duplicates(df1)
df2.head()

중복된 행의 수: 0


,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished


## #04. 결측치 점검

### 1. 결측치 확인

In [6]:
my_qtcheck.check_missing_values(df2)

,Missing Count,Missing Ratio (%)
price,0,0.000
area,0,0.000
bedrooms,0,0.000
bathrooms,0,0.000
stories,0,0.000
mainroad,0,0.000
guestroom,0,0.000
basement,0,0.000
hotwaterheating,0,0.000
airconditioning,0,0.000


## #05. 품질 검사 결과 저장

### 1. 품질 검사 데이터셋 저장

In [7]:
df2.to_excel("housing_qtcheck.xlsx", index=False)

## #06. 기술 통계량 확인

### 1. 명목형 변수의 기술 통계량 확인

In [8]:
category_desc = my_qtcheck.categorical_summary(df2, save_path="housing_category_summary.xlsx")
category_desc

📊 컬럼 'mainroad'의 value_counts():


,count
mainroad,
no,77
yes,468


📊 컬럼 'guestroom'의 value_counts():


,count
guestroom,
no,448
yes,97


📊 컬럼 'basement'의 value_counts():


,count
basement,
no,354
yes,191


📊 컬럼 'hotwaterheating'의 value_counts():


,count
hotwaterheating,
no,520
yes,25


📊 컬럼 'airconditioning'의 value_counts():


,count
airconditioning,
no,373
yes,172


📊 컬럼 'prefarea'의 value_counts():


,count
prefarea,
no,417
yes,128


📊 컬럼 'furnishingstatus'의 value_counts():


,count
furnishingstatus,
furnished,140
semi-furnished,227
unfurnished,178


,mainroad,guestroom,basement,hotwaterheating,airconditioning,prefarea,furnishingstatus
count,545,545,545,545,545,545,545
unique,2,2,2,2,2,2,3
top,yes,no,no,no,no,no,semi-furnished
freq,468,448,354,520,373,417,227


### 2. 연속형 변수의 기술 통계량 확인

In [9]:
desc_df = my_qtcheck.numerical_summary(df2, 
               save_path="housing_numerical_summary.xlsx")
desc_df.T

,price,area,bedrooms,bathrooms,stories,parking
count,545.000,545.000,545.000,545.000,545.000,545.000
mean,4766729.248,5150.541,2.965,1.286,1.806,0.694
std,1870439.616,2170.141,0.738,0.502,0.867,0.862
min,1750000.000,1650.000,1.000,1.000,1.000,0.000
25%,3430000.000,3600.000,2.000,1.000,1.000,0.000
50%,4340000.000,4600.000,3.000,1.000,2.000,0.000
75%,5740000.000,6360.000,3.000,2.000,2.000,1.000
max,13300000.000,16200.000,6.000,4.000,4.000,3.000
rel_diff,0.098,0.120,0.012,0.286,0.097,inf
rdiff_flag,similar,diff,similar,diff,similar,large_diff


## 💡 인사이트

### Q1. 수치/범주 비율은?

- 전체 변수는 **13개**이며, **수치형 6개(46.2%)**, **범주형 7개(53.8%)** 이다.
- 수치형: `price`, `area`, `bedrooms`, `bathrooms`, `stories`, `parking`
- 범주형: `mainroad`, `guestroom`, `basement`, `hotwaterheating`, `airconditioning`, `prefarea`, `furnishingstatus`
- `yes/no` 컬럼은 의미상 수치가 아니라 **이진 범주형 변수**이므로, 품질 점검 단계에서는 **`category` 타입으로 변환**하는 것이 적절하다.
- 이후 회귀/머신러닝 모델링 단계에서는 필요에 따라 **`yes=1, no=0`** 또는 더미변수로 인코딩하면 된다.

### Q2. 결측·중복 있나?

- `info()`와 결측치 점검 결과, **모든 컬럼의 결측치는 0건**이다.
- `duplicated()` 점검 결과, **중복 행은 0건**이다.
- 따라서 이 데이터셋은 **결측치 처리와 중복 제거가 필요하지 않다.**

### Q3. `price` 분포 형태는?

- `price`의 평균은 **4,766,729**, 중앙값은 **4,340,000**으로 평균이 더 크다.
- `area`의 평균은 **5,150.5**, 중앙값은 **4,600**으로 역시 평균이 더 크다.
- 왜도(skew)는 `price = 1.212`, `area = 1.321`로 둘 다 **우편향(right-skewed)** 분포이다.
- 따라서 **`price`와 `area`는 로그 변환(`log1p`) 우선 검토 대상**이다.
- 추가로 모듈 기준에서는 `bathrooms`, `stories`도 로그 변환 후보로 판정되지만, 핵심 대상 변수는 **`price`와 `area`**이다.

### Q4. 이상치 경계는?

- `price`
  - 하한 = **-35,000**
  - 상한 = **9,205,000**
  - 상한 초과 이상치 = **15건 (2.75%)**
  - 하한 미만 이상치 = **0건 (0.00%)**

- `area`
  - 하한 = **-540**
  - 상한 = **10,500**
  - 상한 초과 이상치 = **12건 (2.20%)**
  - 하한 미만 이상치 = **0건 (0.00%)**

- `bedrooms`
  - 하한 = **0.5**
  - 상한 = **4.5**
  - 상한 초과 이상치 = **12건 (2.20%)**
  - 하한 미만 이상치 = **0건 (0.00%)**

- `bathrooms`
  - 하한 = **-0.5**
  - 상한 = **3.5**
  - 상한 초과 이상치 = **1건 (0.18%)**
  - 하한 미만 이상치 = **0건 (0.00%)**

- `stories`
  - 하한 = **-0.5**
  - 상한 = **3.5**
  - 상한 초과 이상치 = **41건 (7.52%)**
  - 하한 미만 이상치 = **0건 (0.00%)**

- `parking`
  - 하한 = **-1.5**
  - 상한 = **2.5**
  - 상한 초과 이상치 = **12건 (2.20%)**
  - 하한 미만 이상치 = **0건 (0.00%)**


### 질문 내용을 통해 도출한 결론

#### Step 1·2

- `price`와 `area`는 평균 > 중앙값, 왜도 > 1로 확인되어 **우편향 분포**이다.
- 따라서 두 변수는 **로그 변환 우선 대상**으로 식별된다.

#### Step 3·4

- 변동계수(CV)는 `area = 0.421`, `price = 0.392`로, **`area`의 상대 변동성이 약간 더 크다.**
- IQR 기준 상한 초과 비율은 `area 2.20%`, `price 2.75%`로 확인되었다.

#### Step 5

- 범주형 시설 변수는 가격 프리미엄과 관련이 있다.
  - `airconditioning = yes`는 `no` 대비 평균 가격이 **약 43.45% 높다.**
  - `prefarea = yes`는 `no` 대비 평균 가격이 **약 32.85% 높다.**
  - `hotwaterheating = yes`는 `no` 대비 평균 가격이 **약 17.58% 높다.**
  - `furnishingstatus`는 **furnished > semi-furnished > unfurnished** 순으로 평균 가격이 높다.
- 즉, **시설 보유 여부와 선호지역 여부는 가격 프리미엄 형성에 유의미한 영향을 주는 변수**로 볼 수 있다.